File: containerWithMostWater/c++/Container_with_most_water.cpp
//
//  Container_with_most_water.cpp
//  
//
//  Created by Wen-Ting Wang on 2018/12/6.
//

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0, right = height.size() - 1, area = 0;
        
        while(left < right){
            area = max(area, min(height[left], height[right]) * (right - left));
            height[left] < height[right] ? ++ left : -- right;
        }
        return area;
    }
};


int main() {
    int array[] = {1,8,6,2,5,4,8,3,7};
    vector<int> height;
    height.assign(array, array+9);
    
    int ret = Solution().maxArea(height);
    
    cout << ret << endl;
    return 0;
}

File: problems/containerWithMostWater/python/Container_with_most_water.py
class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """

        left, right, area = 0, len(height) - 1, 0

        while left < right:

            minHeight = min(height[left], height[right])
            area = max(area, minHeight * (right - left))

            while left < right and minHeight == height[left]:
                left += 1
            while left < right and minHeight == height[right]:
                right -= 1

        return area


if __name__ == "__main__":
    height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    print(Solution().maxArea(height))
